src/serializer/visitors.js

```diff
@@ -98,13 +99,25 @@ export let ClosureRefReplacer = {
   },
 };
 
+// todo: make a test case that throws a FatalError when calling this. First try a residual function with a reference
+// to a non existant global.
+
 function visitName(state, name, modified) {
   let doesNotMatter = true;
-  let ref = state.tryQuery(
-    () => ResolveBinding(state.realm, name, doesNotMatter, state.val.$Environment),
-    undefined,
-    true
-  );
+  let oldErrorHandler = state.realm.errorHandler;
+  state.realm.errorHandler = d => "Fail";
+  let ref;
+  try {
+    ref = state.tryQuery(
+      () => ResolveBinding(state.realm, name, doesNotMatter, state.val.$Environment),
+      undefined,
+      true
+    );
+  } catch (e) {
+    if (!(e instanceof FatalError)) throw e;
```

Comment:
`tryQuery` is used from a few more places. I think the whole error-handling logic should move into `tryQuery`.